evm/governance.js

```diff
@@ -0,0 +1,285 @@
+'use strict';
+
+require('dotenv').config();
+
+const {
+    Wallet,
+    getDefaultProvider,
+    utils: { isAddress, defaultAbiCoder, keccak256, Interface },
+    Contract,
+    BigNumber,
+} = require('ethers');
+const readlineSync = require('readline-sync');
+const { Command, Option } = require('commander');
+
+const {
+    printInfo,
+    printWalletInfo,
+    loadConfig,
+    isNumber,
+    isValidTimeFormat,
+    etaToUnixTimestamp,
+    getCurrentTimeInSeconds,
+    wasEventEmitted,
+    printWarn,
+    printError,
+} = require('./utils');
+const IGovernance = require('@axelar-network/axelar-gmp-sdk-solidity/interfaces/IAxelarServiceGovernance.json');
+
+async function processCommand(options, chain) {
+    const { contractName, address, governanceAction, calldata, nativeValue, eta, privateKey, yes } = options;
+
+    if (contractName !== 'AxelarServiceGovernance' && contractName !== 'InterchainGovernance') {
+        throw new Error(`Invalid governance contract: ${contractName}`);
+    }
+
+    const contracts = chain.contracts;
+    const contractConfig = contracts[contractName];
+
+    let governanceAddress;
+
+    if (isAddress(address)) {
+        governanceAddress = address;
+    } else {
+        if (!contractConfig?.address) {
+            throw new Error(`Contract ${contractName} is not deployed on ${chain.name}`);
+        }
+
+        governanceAddress = contractConfig.address;
+    }
+
+    const target = chain.contracts.AxelarGateway?.address;
+
+    if (!isAddress(target)) {
+        throw new Error(`Missing AxelarGateway address in the chain info.`);
+    }
+
+    if (!isNumber(nativeValue)) {
+        throw new Error(`Invalid native value: ${nativeValue}`);
+    }
+
+    if (!isValidTimeFormat(eta)) {
+        throw new Error(`Invalid ETA: ${eta}. Please pass the eta in the format YYYY-MM-DDTHH:mm:ss`);
+    }
+
+    const rpc = chain.rpc;
+    const provider = getDefaultProvider(rpc);
+
+    const wallet = new Wallet(privateKey, provider);
+    await printWalletInfo(wallet);
+
+    printInfo('Contract name', contractName);
+
+    const governanceContract = new Contract(governanceAddress, IGovernance.abi, wallet);
+
+    const gasOptions = contractConfig.gasOptions || chain.gasOptions || {};
+    console.log(`Gas override for chain ${chain.name}: ${JSON.stringify(gasOptions)}`);
+
+    printInfo('Proposal Action', governanceAction);
+
+    const unixEta = etaToUnixTimestamp(eta);
+
+    const types = ['uint256', 'address', 'bytes', 'uint256', 'uint256'];
+    const values = [0, target, calldata, nativeValue, unixEta];
+
+    let gmpPayload;
+
+    switch (governanceAction) {
+        case 'scheduleTimeLock': {
```

Comment:
Shouldn't we also call the `execute()` function for the first 4 switch cases, as to actually process the specific `ServiceGovernanceCommand` command, or may be i am missing something.

Reply:
The external execute function is in AxelarExecutable and it checks for gateway approval before calling the internal execute function so calling it as part of the switch case would fail as the call will not yet have been approved by the gateway.